File: report/src/main/java/com/vmlens/report/runelementtype/MethodWithLockAccess.java
package com.vmlens.report.runelementtype;

import com.vmlens.report.description.DescriptionContext;
import com.vmlens.report.description.NeedsDescriptionCallback;

public class MethodWithLockAccess implements RunElementType {

    private final int atomicMethodId;
    private final long objectHashCode;
    private final LockOperation lockOperation;
    private final ReportLockType lockType;

    public MethodWithLockAccess(int atomicMethodId,
                                long objectHashCode, 
                                LockOperation lockOperation,
                                ReportLockType lockType) {
        this.atomicMethodId = atomicMethodId;
        this.objectHashCode = objectHashCode;
        this.lockOperation = lockOperation;
        this.lockType = lockType;
    }

    @Override
    public String operation() {
        return  lockType.textForMethodWithLock() + " " + lockOperation.text();
    }

    @Override
    public String element(DescriptionContext context) {
        return context.methodName(atomicMethodId) + "@" + objectHashCode;
    }

    @Override
    public void addToNeedsDescription(NeedsDescriptionCallback callback) {
        callback.needsMethod(atomicMethodId);
    }
}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/runelementtype/MethodWithLockAccess.scala
package com.anarsoft.race.detection.report.element.runelementtype

import com.anarsoft.race.detection.report.description.{DescriptionContext, NeedsDescriptionCallback}
import com.anarsoft.race.detection.report.element.runelementtype.objecthashcodemap.ObjectHashCodeMap


class MethodWithLockAccess(val atomicMethodId: Int,val objectHashCode: Long, val lockOperation: LockOperation,val lockType: ReportLockType) extends ReportOperation {
  private var objectHashCodeMap: ObjectHashCodeMap = null

  override def operation: String = lockType.textForMethodWithLock + " " + lockOperation.text

  override def element(context: DescriptionContext): String = context.methodName(atomicMethodId)

  override def addToNeedsDescription(callback: NeedsDescriptionCallback): Unit = {
    callback.needsMethod(atomicMethodId)
  }

  override def objectString(context: DescriptionContext): String = objectHashCodeMap.asUiString(objectHashCode)

  override def setObjectHashCodeMap(objectHashCodeMap: ObjectHashCodeMap, threadIndex: Int): Unit = {
    this.objectHashCodeMap = objectHashCodeMap
    objectHashCodeMap.add(objectHashCode, threadIndex)
  }
}
